### PR TITLE
Added Buffer2 class

### PR DESCRIPTION
**Pull request by lucaswoj**
**recut from #1414**

Up until now, GL JS has used hand-coded classes to specify buffer layouts and buffer manipulation operations. This worked well pre-data-driven-styling because there were < 10 possible buffer layouts. For example, [here is the `TriangleElementBuffer` class](https://github.com/mapbox/mapbox-gl-js/blob/master/js/data/buffer/triangle_element_buffer.js)

```js
module.exports = TriangleElementBuffer;

function TriangleElementBuffer(buffer) {
    Buffer.call(this, buffer);
}

TriangleElementBuffer.prototype = util.inherit(Buffer, {
    itemSize: 6, // bytes per triangle (3 * unsigned short == 6 bytes)
    arrayType: 'ELEMENT_ARRAY_BUFFER',

    add: function(a, b, c) {
        var pos2 = this.pos / 2;

        this.resize();

        this.ushorts[pos2 + 0] = a;
        this.ushorts[pos2 + 1] = b;
        this.ushorts[pos2 + 2] = c;

        this.pos += this.itemSize;
    }
});
```

Post-data-driven styling, however, the number of possible buffer layouts is [practically infinite](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) because any subset of paint properties might be included in the buffer. 

The first step towards proper data driven styling was to create a more general and powerful framework for buffers (temporarily called `Buffer2`) which allows buffer layouts to be specified by configuration rather than code. Using `Buffer2`, We could write the same `TriangleElementBuffer` class as

```js
module.exports = function() {
    Buffer2.apply(this, {
        type: Buffer2.BufferType.ELEMENT,
        attributes: {
            verticies: {
                components: 3,
                type: Buffer2.AttributeType.UNSIGNED_SHORT
            }
        }
    });
}
```

... but in practice we *wouldn't* write it that way. GL JS would choose the attributes for that buffer at runtime for the best performance. 

The `Buffer2` class will know how to construct, write to, and read from GL buffers but have no semantic understanding of their contents. 

Instances of `Buffer2` can be ([destructively](https://developers.google.com/web/updates/2011/12/Transferable-Objects-Lightning-Fast?hl=en)) serialized and transferred across threads. 

cc @tmcw @mourner @kkaefer

### CODE REVIEW

**Comment by mourner** [comment]
buffer_shim stuff is temporary until you move over everything to Buffer2, right?

File: js/data/buffer2.js [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        this.size = clone.size;
+        this.length = clone.index;
+        this.arrayBuffer = clone.arrayBuffer;
+        this.attributes = clone.attributes;
+        this.itemSize = clone.itemSize;
+
+        this._refreshArrayBufferViews();
+    }
+
+    util.assert(this.type);
```
Comment (mcwhittemore) [inline js/data/buffer2.js]: @lucaswoj wanna move this to the top of the function so we don't do the above work if we're just going to throw an error.

Comment (mourner) [inline js/data/buffer2.js]: well, if it we're actually throwing an error, performance doesn't matter much :)

Comment (mcwhittemore) [inline js/data/buffer2.js]: true, still, better to do validation before the work not after, if only for readability :)

**Comment by lucaswoj** [comment]
@mourner Yes, the "shim" is a temporary measure to DRY up the code that wraps a `Buffer2` in an old-style "`Buffer1`" per your suggestion in https://github.com/mapbox/mapbox-gl-js/pull/1414#issuecomment-129908062. 

My intention is for the shim to get merged into master temporarily so we can start testing `Buffer2` in the real world before we finish rewriting everything that touches `Buffer1`.

**Comment by mourner** [comment]
Are they incompatible enough to write a shim? I see that it proxies most of the methods. Could you just make Buffer2 have things that other code relies on in Buffer1, and then remove any cruft after rewriting that code?

File: js/data/buffer2.js (hunk 1 above)
Comment (lucaswoj) [inline js/data/buffer2.js]: This is an overloaded constructor. Like `Buffer1`, it can either

 - create an empty buffer from a configuration object
 - rehydrate a serialized instance of `Buffer`

The idea with the `util.assert(this.type)` line is that we allow each case to parse the constructor arguments independently and then perform a little validation on the resulting instance. It is not necessary to perform this assertion but I found that `this.type` had a tendency to end up as `undefined` and that resulted in hard-to-debug errors.

In a larger sense, I sorta want to rewrite this whole constructor to eliminate this branching altogether. It should be possible to combine these two cases into one.

Comment (mourner) [inline js/data/buffer2.js]: I'd also consider removing asserts once the code has been tested and finalized. It's kinda all or nothing — either we add asserts everywhere throughout the whole code bringing a lot of cruft with it, or don't add them and rely on tests and throwing specific errors in special cases instead.

**Comment by lucaswoj** [comment]
The latest commit

 - removes the shim in favor of direct compatibility with `Buffer1`
 - removes the special serialization logic (which was both unnecessary and a huge pain to support)
 - implements `TriangleElementBuffer` as a subclass of `Buffer2`

File: js/data/buffer/triangle_element_buffer.js [hunk 2]
```diff
@@ -1,27 +1,24 @@
...
+
+function TriangleElementBuffer(options) {
+    Buffer2.call(this, options || {
+        type: Buffer2.BufferType.ELEMENT,
+        attributes: {
+            verticies: {
+                components: 3,
+                type: Buffer2.AttributeType.UNSIGNED_SHORT
+            }
+        }
```
Comment (mourner) [inline js/data/buffer/triangle_element_buffer.js]: could those move to the prototype below?

File: test/js/buffer/triangle_element_buffer.test.js [hunk 3]
```diff
@@ -1,11 +1,10 @@
 'use strict';
 
 var test = require('prova');
-var Buffer = require('../../../js/data/buffer/buffer');
 var TriangleElementBuffer = require('../../../js/data/buffer/triangle_element_buffer');
 
 test('TriangleElementBuffer', function(t) {
-    var buf = new Buffer();
+    var buf = new TriangleElementBuffer();
     t.ok(new TriangleElementBuffer(buf), 'default buffer');
```
Comment (mourner) [inline test/js/buffer/triangle_element_buffer.test.js]: double construction?

**Comment by mourner** [comment]
@lucaswoj looks great! Let's figure out the fill issues before proceeding.

File: js/util/util.js [hunk 4]
```diff
@@ -400,3 +400,15 @@ exports.getCoordinatesCenter = function(coords) {
...
+ * Throw an error if a condition is not met
+ * @param {boolean} condition
+ * @returns {?string} message
+ * @private
+ */
+exports.assert = function(condition, message) {
+    if (!condition) {
+        throw new Error(message || 'Assertion failed');
+    }
+};
```
Comment (mourner) [inline js/util/util.js]: Could we just use `console.assert` instead of a custom function here? In addition to making it unnecessary, this has an advantage of being supported by https://github.com/twada/unassertify

File: js/data/buffer/triangle_element_buffer.js (hunk 2 above)
Comment (lucaswoj) [inline js/data/buffer/triangle_element_buffer.js]: Yes, the attributes could move to the prototype, but ideally they wouldn't move to the prototype. The raison-d'etre of `Buffer2` is that it has a configurable buffer layout (vs `Buffer1` which had a hardcoded buffer layout). Layout is intended to be passed as a constructor argument at runtime. This code represents an intermediary step towards that goal.

File: test/js/buffer/triangle_element_buffer.test.js (hunk 3 above)
Comment (lucaswoj) [inline test/js/buffer/triangle_element_buffer.test.js]: This is a strange test but rewriting it feels out of scope for this PR. `Buffer2` has [good test coverage](https://github.com/mapbox/mapbox-gl-js/pull/1526/files#diff-4e94531e9a19c46e6d705f5a53c60dabR1).

File: js/util/util.js (hunk 4 above)
Comment (lucaswoj) [inline js/util/util.js]: Looks like `console.assert` is not and will not be widely supported. `util.assert` is just a shorthand around `if (blah) throw new Error(...)`. Validating preconditions is especially helpful because WebGL is bad at reporting errors. I can remove the `util.assert` statements before merging this code if you prefer.

**Comment by lucaswoj** [comment]
Latest commit addresses PR feedback, renames `Buffer2::size` to `Buffer2::capacity` (a la c++ vectors) and fixes rendering issues.

File: js/util/util.js (hunk 4 above)
Comment (lucaswoj) [inline js/util/util.js]: ref https://github.com/mapbox/mapbox-gl-js/issues/1543#issuecomment-143013921

**Comment by lucaswoj** [comment]
FPS benchmarks look :+1: so far 

```
duration: 3000
29 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.10.0/mapbox-gl.js
29 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.11.0/mapbox-gl.js
29.2 fps /debug/mapbox-gl.js
28.7 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.10.0/mapbox-gl.js
28.9 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.11.0/mapbox-gl.js
29 fps /debug/mapbox-gl.js
```

File: js/util/util.js (hunk 4 above)
Comment (mourner) [inline js/util/util.js]: No, I'm convinced on assertions — let's keep, but looks like we'll use the `assert` module so that it works with unassertify.

Comment (jfirebaugh) [inline js/util/util.js]: Just merged #1557, so if you rebase on master you can start using `assert`.

**Comment by jfirebaugh** [comment]
I would expect performance effects to be largely confined to the Worker. The end result is the same on the GPU, so FPS wouldn't change, but the time taken to build the Buffer in the background could change. It would be nice to have some benchmarks for that.

**Comment by lucaswoj** [comment]
@jfirebaugh Sounds good. How about we add a benchmark that measures the total time spent creating buffers during a slow `flyTo` from DC to SF? If that sounds :+1: I'll create a ticket to implement that bench script.

**Comment by lucaswoj** [comment]
A *very* unscientific test using the new `bufferstats` branch seems to indicate that `Buffer2` does not add substantial overhead to buffer creation.

**Comment by lucaswoj** [comment]
Thanks to @mcwhittemore, we now have all buffers rewritten using Buffer2 and...

*drumroll*...

Its about 10x slower than master

```
## buffer2 
duration: 30000
9140 ms, 46 samples for /debug/mapbox-gl.js
10198 ms, 45 samples for /debug/mapbox-gl.js

# master
duration: 30000
1225 ms, 112 samples for /debug/mapbox-gl.js
1235 ms, 119 samples for /debug/mapbox-gl.js
```

I haven't even begun to think about performance optimizations, so I'm hopeful that there'll be some easy gains.

**Comment by lucaswoj** [comment]
Got the `bucket-stats` benchmark down to ~3500 ms in the latest push. I'm not seeing any more low hanging fruit right now. 

```
3654 ms, 111 samples for /debug/mapbox-gl.js
3416 ms, 116 samples for /debug/mapbox-gl.js
```

**Comment by mourner** [comment]
Just as I predicted a while back. I could take a look at optimizing this more. 

In the worst case, if we can't make any more perf improvements and it's still too slow, there's an option of code generation, similar to what I did with layer filters. It can bring us back to master performance while being dynamically generated from attribute config.

File: js/data/buffer.js [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+/**
+ * Set an item at a particular index. Grows the buffer if necessary.
+ * @private
+ * @param {number} index The index of the item to set
+ * @param {(BufferItem|BufferValue)} item the item to set. If the buffer only has one attribute,
+ * it can be a single value instead of an item.
+ */
+Buffer.prototype.set = function(index, item) {
+    if (!Array.isArray(item)) {
```
Comment (jfirebaugh) [inline js/data/buffer.js]: Performance optimization: eliminate this condition. Either:

* Always require an array, or
* Write `this.set = function() { ... }` dynamically in the constructor, and make it require an array for multiple attributes, and require no array for single attributes. For multiple attributes, it may help to iterate over `options.attributes` rather than `item` (like, maybe the compiler will detect that `options.attributes` is constant).

Writing `this.set = function() { ... }` dynamically in the constructor can hopefully give you some nice speed ups, while stopping just short of needing code-gen + `eval`.

File: js/data/buffer/line_vertex_buffer.js [hunk 6]
```diff
@@ -1,46 +1,46 @@
...
-            index = this.index,
-            extrudeScale = LineVertexBuffer.extrudeScale;
-
-        this.resize();
-
-        this.shorts[pos2 + 0] = (Math.floor(point.x) * 2) | tx;
-        this.shorts[pos2 + 1] = (Math.floor(point.y) * 2) | ty;
-
-        this.bytes[pos + 4] = Math.round(extrudeScale * extrude.x);
-        this.bytes[pos + 5] = Math.round(extrudeScale * extrude.y);
```
Comment (lucaswoj) [inline js/data/buffer/line_vertex_buffer.js]: `LineVertexBuffer::add` is one of the hottest functions in the codebase. It looks like dropping this call to `Math.round` gives us a substantial performance boost but changes the output *just* enough to break test suite. Can we afford to modify test-suite for this? Can we leverage this information in another way? cc @mourner @jfirebaugh

Comment (jfirebaugh) [inline js/data/buffer/line_vertex_buffer.js]: cc @kkaefer -- the rounding here goes _way_ back, at least as far as https://github.com/mapbox/mapbox-gl-js/commit/5cf0c0b513cd851ad4d91bd238760c34f117c072#diff-ede55fc20965ad828d7e5cf6eaa5c404R50.

**Comment by lucaswoj** [comment]
Just wrote up a little code-gen prototype. With the removal of `Math.round` thrown in, the `buffer2` branch is neck and neck with `master`!

```
1751 ms, 127 samples for /debug/mapbox-gl.js
1846 ms, 123 samples for http://localhost:9000/mapbox-gl-master.js
2440 ms, 115 samples for /debug/mapbox-gl.js
2225 ms, 120 samples for http://localhost:9000/mapbox-gl-master.js
```

**Comment by lucaswoj** [comment]
... without the removal of `Math.round`, things don't look so great

```
duration: 30000
2355 ms, 122 samples for /debug/mapbox-gl.js
1856 ms, 130 samples for http://localhost:9000/mapbox-gl-master.js
3302 ms, 115 samples for /debug/mapbox-gl.js
1999 ms, 124 samples for http://localhost:9000/mapbox-gl-master.js
```

I'm going to open a separate ticket about `Math.round`

**Comment by mourner** [comment]
@lucaswoj if the methods are generated now, why is it slower with `Math.round` if it was present before? Performance should be the same if we generate the same code.

**Comment by mourner** [comment]
One of the things that could cause slowdowns is the need to create new objects and arrays for every push, and also iterate with `for in` which is slow too. Now we have the following syntax:

```js
this.push({
    pos: [
        (point.x << 1) | tx,
        (point.y << 1) | ty
    ],
    data: [
        EXTRUDE_SCALE * extrude.x,
        EXTRUDE_SCALE * extrude.y,
        Math.round(EXTRUDE_SCALE * extrude.x),
        Math.round(EXTRUDE_SCALE * extrude.y)
    ]
});
```

Can we change the generated code to turn this into e.g. the following?

```js
this.push('pos', 
    (point.x << 1) | tx, 
    (point.y << 1) | ty);

this.push('data', EXTRUDE_SCALE * extrude.x,
    EXTRUDE_SCALE * extrude.y,
    Math.round(EXTRUDE_SCALE * extrude.x),
    Math.round(EXTRUDE_SCALE * extrude.y));
```

**Comment by mourner** [comment]
Yep, that seems to do it. Sketched out the idea in https://github.com/mapbox/mapbox-gl-js/commit/bea8d6dccf176c88394de8badde5abf1c7f815c5. On my machine, the bench produces:

```js
539 ms, 128 samples for /dist/mapbox-gl.js # master
741 ms, 128 samples for /dist/mapbox-gl.js # buffer2
566 ms, 128 samples for /dist/mapbox-gl.js # flat-push
558 ms, 129 samples for /dist/mapbox-gl.js # flat-push, no Math.round
```

For some reason removing `Math.round` doesn't affect my results much, but this may be machine-specific.

Also note how the buffer code could be simplified further with this flat approach.

**Comment by mourner** [comment]
On a side note, just noticed that `glyph_vertex_buffer` and `icon_vertex_buffer` are pretty much identical. We can remove one of them.

File: js/data/buffer/circle_vertex_buffer.js [hunk 7]
```diff
@@ -1,36 +1,26 @@
...
-    Buffer.call(this, buffer);
+var Buffer = require('../buffer');
+
+function CircleVertexBuffer(options) {
+    Buffer.call(this, options || {
+        type: Buffer.BufferType.VERTEX,
+        attributes: [{
+            name: 'pos',
+            components: 2,
+            type: Buffer.AttributeType.UNSIGNED_SHORT
```
Comment (mourner) [inline js/data/buffer/circle_vertex_buffer.js]: This should be `SHORT` rather than `UNSIGNED_SHORT`.

File: js/data/buffer/fill_vertex_buffer.js [hunk 8]
```diff
@@ -1,25 +1,23 @@
...
-    Buffer.call(this, buffer);
+var Buffer = require('../buffer');
+
+function FillVertexBuffer(options) {
+    Buffer.call(this, options || {
+        type: Buffer.BufferType.VERTEX,
+        attributes: [{
+            name: 'pos',
+            components: 2,
+            type: Buffer.AttributeType.UNSIGNED_SHORT
```
Comment (mourner) [inline js/data/buffer/fill_vertex_buffer.js]: This should be `SHORT` rather than `UNSIGNED_SHORT` as well.

**Comment by mourner** [comment]
Also, `line_element_buffer` is identical to `triangle_element_buffer` so can be removed as well.

Generally I see that we're in a great position to remove `*_buffer.js` classes in this PR and turn them into sets of attribute configs like @lucaswoj envisioned (even before writing `Layer`). For that to happen, we need to eliminate custom `add` and `bind` methods.

To eliminate `add` (that's already identical to `push` in my branch for many buffer types), we simply move the packing logic to the corresponding bucket implementations, so this:

```js
add: function(x, y, extrudeX, extrudeY) {
    this.push(
        (x * 2) + ((extrudeX + 1) / 2),
        (y * 2) + ((extrudeY + 1) / 2));
},
...
circleVertex.add(x, y, -1, -1);
```

Becomes:

```js
circleVertex.push(x * 2, y * 2);
```

We also may want to keep the multiply-floor packing logic like this in the attribute configs:

```js
this.push(
    x, y,
    Math.floor(ox * 64), // use 1/64 pixels for placement
    Math.floor(oy * 64),
    Math.floor(tx / 4), /* tex */
    Math.floor(ty / 4), /* tex */
```

To do this, we simply add a `multiplier` option to attributes so this becomes:

```js
this.push(
    x, y,
    ox, oy,
    tx, ty,
...
attributes: [{
    name: 'pos',
    components: 2,
    type: Buffer.AttributeType.SHORT
}, {
    name: 'extrude',
    components: 2,
    type: Buffer.AttributeType.SHORT,
    multiplier: 64, // use 1/64 pixels for placement
}, {
    name: 'data1',
    components: 4,
    type: Buffer.AttributeType.UNSIGNED_BYTE,
    multiplier: 1 / 4
}, {
```

To eliminate custom `bind`, we can simply write a generic implementation that uses the attributes configuration of the buffer. Currently this kind of code is redundant:

```js
gl.vertexAttribPointer(shader.a_pos, 2, gl.SHORT, false, stride, offset + 0);
gl.vertexAttribPointer(shader.a_offset, 2, gl.SHORT, false, stride, offset + 4);
gl.vertexAttribPointer(shader.a_data1, 4, gl.UNSIGNED_BYTE, false, stride, offset + 8);
gl.vertexAttribPointer(shader.a_data2, 2, gl.UNSIGNED_BYTE, false, stride, offset + 12);
```

To make things more confusing, vertexAttribPointer configuration happens in two places. For some buffer types like line vertex, it's inside `draw_*.js`, and for others like icon vertex it's in `*_buffer.js`. So writing a generic `bind` in `buffer.js` will make the code much cleaner.

**Comment by mourner** [comment]
Feel free to fast-forward `flat-push` branch into this one if the changes make sense.

**Comment by lucaswoj** [comment]
Thanks for the :eyes: @mourner! I appreciate your thoughtful feedback. Below, I've enumerated your proposed next steps and added my own thoughts.

### add a `multiplier` option to attributes

I agree that this multiplier should be pulled out into configuration :+1:. However, I'm envisioning it living on the `StyleLayer2` class, where it can be shared between buffer attribute generation and constant attribute generation (i.e. with and without `disableVertexAttribArray`). I propose we leave it inlined with calls to `add` for now and pull it out in the `StyleLayer2` PR. 

## merge `flat-push` branch

Sad to lose the self-documenting object keys but this will make hot code faster so we should do it. :white_check_mark: 

## remove all `*_buffer` classes

 - add a universal `bind` method
 - remove `add` method

The plan was to land this in a separate PR but it looks like a small undertaking with a big performance bump, so we should do it! :white_check_mark: 

## drop `Math.round` in all buffers

Let's do this one in a separate PR. It is conceptually unrelated and might generate some discussion. ref #1573 :white_check_mark: 

## merge `line_element_buffer` and `triangle_element_buffer` classes

This will be subsumed by "remove all `*_buffer` classes"

**Comment by mourner** [comment]
@lucaswoj good plan, let's move with this!

> The plan was to land this in a separate PR but it looks like a small undertaking with a big performance bump, so we should do it!

Actually I think we can leave this for a separate PR. It probably doesn't affect performance too much, I was looking forward to it mostly because it removes unnecessary code and keeps all buffer logic in one neat place.

> merge line_element_buffer and triangle_element_buffer classes

Also `icon_vertex_buffer` vs `glyph_vertex_buffer`.

**Comment by lucaswoj** [comment]
I think this is ready to :ship:! Any last words?

```
duration: 30000
1649 ms, 116 samples for /debug/mapbox-gl.js
1678 ms, 124 samples for http://localhost:9000/mapbox-gl-master.js
2846 ms, 106 samples for /debug/mapbox-gl.js
2210 ms, 121 samples for http://localhost:9000/mapbox-gl-master.js
```

cc @mourner @jfirebaugh @mcwhittemore

File: js/data/buffer.js [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        SHORT:          new Int16Array(this.arrayBuffer)
+    };
+};
+
+Buffer.prototype._createPushMethod = function() {
+    var body = '';
+
+    body += 'var index = this.length++;\n';
+    body += 'var offset = index * ' + this.itemSize + ';\n';
+    body += 'if (offset + ' + this.itemSize + ' > this.capacity) { this._resize(this.capacity * 1.5); }\n';
```
Comment (mcwhittemore) [inline js/data/buffer.js]: itemSize is a static value? @lucaswoj

Comment (mcwhittemore) [inline js/data/buffer.js]: NVM. Offset is what's changing here.

File: js/data/buffer.js [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+Buffer.prototype._refreshArrayBufferViews = function() {
+    this.arrayBufferViews = {
+        UNSIGNED_BYTE:  new Uint8Array(this.arrayBuffer),
+        BYTE:           new Int8Array(this.arrayBuffer),
+        UNSIGNED_SHORT: new Uint16Array(this.arrayBuffer),
+        SHORT:          new Int16Array(this.arrayBuffer)
+    };
+};
+
+Buffer.prototype._createPushMethod = function() {
```
Comment (mcwhittemore) [inline js/data/buffer.js]: Why do this as its own function and not just part of the Buffer function? Doing it this ways adds both the _createPushMethod and the attributeNames to this?

Comment (mourner) [inline js/data/buffer.js]: Can you clarify the question? Not sure what you're worried about here...

Comment (lucaswoj) [inline js/data/buffer.js]: @mcwhittem I don't love cluttering the `Buffer` prototype with this method... It felt like too many LOC to shove this logic in the constructor and too entangled with the object state to break out into a function.

**Comment by lucaswoj** [comment]
Going to :ship: as soon as CI passes

**Comment by mourner** [comment]
Not sure why you squashed everything into one commit, both me and John where not in favor of that...

**Comment by mourner** [comment]
Oh, sorry, I see the problem now — it was hard to rebase in a way where each commit is in a good state, so squashing is good here. Fair enough.

**Comment by lucaswoj** [comment]
Sorry about the squash, @mourner. I tried to come up with a clean rebase but there were a bugs and accidental check-ins in the intermediate commits.
